raidtools2-0.90.20010914/mkpv.8
.\" -*- nroff -*-
.TH mkpv 8
.SH NAME
mkpv \- creates PV
.SH SYNOPSIS
\fBmkpv\fR [--configfile] [--version] [--force]
	     [--upgrade] [-acvfuv] </dev/md?>+


.SH DESCRIPTION
\fBmkpv\fR sets up a block devices into a single PV.

Note that initializing PV devices \fBdestroys all of the data on
the consituent devices\fR. 

.SH OPTIONS
.IP "\fB-c\fP, \fB\-\-configfile\fP \fIfilename\fR"
Use \fIfilename\fR as the configuration file (\fB/etc/raidtab\fR is used
by default).

.IP "\fB\-f\fP, \fB\-\-force\fP"
Initilize the consituent devices, even if they appear to have data on
them already.

.IP "\fB\-h\fP, \fB\-\-help\fP"
Displays a short usage message, then exits.

.IP "\fB\-V\fP, \fB\-\-version\fP"
Displays a short version message, then exits.

.SH NOTES

The raidtools are derived from the md-tools and raidtools packages, which
were originally written by Marc Zyngier, Miguel de Icaza, Gadi Oxman, 
Bradley Ward Allen, and Ingo Molnar.

.SH AUTHOR
Ingo Molnar

.SH BUGS
Probably many.

.SH SEE ALSO
.IR raidtab (5),
.IR raidstart (8)


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



raidtools2-0.90.20010914/mkpv.c
/*
 * mkpv.c : LVM Utility for the Linux Logical Volume Manger driver
 *          Copyright (C) 1998 Ingo Molnar
 *
 * This utility creates a Physical Volume
 *
 * This source is covered by the GNU GPL, the same as all Linux kernel
 * sources.
 *
 * 2001-12-21: /proc/mount checking, Darik Horn <dajhorn at uwaterloo dot ca>
 *
 */

#include "common.h"
#include "parser.h"
#include "raid_io.h"
#include "raidlib.h"
#include "version.h"
#include <popt.h>
#include <sys/mount.h>		/* for BLKGETSIZE */
#ifndef BLKGETSIZE
#include <linux/fs.h>          /* for BLKGETSIZE */
#endif
#include <sys/sysmacros.h>

#include "lvm-int.h"

#define MKPV_MAJOR_VERSION            (0)
#define MKPV_MINOR_VERSION            (90)
#define MKPV_PATCHLEVEL_VERSION       (0)

void usage (void) {
    printf("usage: mkpv [--configfile] [--version] [--force]\n");
    printf("       [-acfhuv] </dev/md?>*\n");
}

static long long lvmseek (int fd, unsigned long block)
{
	return (raidseek(fd, block*(LVM_BLOCKSIZE/MD_BLK_SIZ)));
}

static int mkvg (int fd, int forceSanity)
{
	unsigned char buffer[LVM_BLOCKSIZE];
	vg_sb_t *vg_sb;
	lv_descriptor_t *lv;

	if (lvmseek(fd, 1) == -1)
		return 1;
	if ((read(fd, buffer, LVM_BLOCKSIZE)) != LVM_BLOCKSIZE)
		return 1;
	vg_sb = (vg_sb_t *) buffer;

	if (!forceSanity) {
		/*
		 * take a look at the superblock ...
		 */
		if (vg_sb->vg_magic == LVM_VG_SB_MAGIC) {
			fprintf(stderr, "VG appears to be already created -- use -f to\nforce the destruction of old contents\n");
			return 1;
		}
	}

	memset(buffer,0,LVM_BLOCKSIZE);

	printf("creating VG ...\n");
	vg_sb->vg_magic = LVM_VG_SB_MAGIC;

	vg_sb->nr_lvs = 1;

	printf("creating LV 1 ... \n");

	lv = vg_sb->lv_array + 1;
	lv->lv_id = 1;
	lv->lv_max_indices = 20000;
	lv->lv_free_indices = lv->lv_max_indices;
	lv->md_id = 9;
	
	if (lvmseek(fd, 1) == -1)
		return 1;
	if ((write(fd, buffer, LVM_BLOCKSIZE)) != LVM_BLOCKSIZE)
		return 1;

	return 0;
}

static int mkpv (char *name, int forceSanity)
{
	FILE *fp;
	unsigned char buffer[LVM_BLOCKSIZE];
	pv_sb_t *pv_sb;
	int fd, i, block_groups, bg_size;
	unsigned long nr_blocks;
	pv_block_group_t *pv_bg;
	struct mntent * mntinfo;
	int err;

	fd = open(name, O_RDWR);
	if (fd == -1) {
		perror("make PV: ");
		return 1;
	}

	/*
	 * Check if the device is mounted
	 */
	if (!(fp = setmntent("/etc/mtab", "r")) && !(fp = setmntent("/proc/mounts", "r"))) {
		fprintf(stderr, "mkpv: unable to open both /etc/mtab and /proc/mounts\n");
		return 1;
	}
	
	err = 0;
	while ((mntinfo = getmntent(fp))) {
	    if (!strcmp(mntinfo->mnt_fsname, name)) {
		    fprintf(stderr, "%s is mounted\n", name);
		    err = 1;
	    }
	}

	endmntent(fp);
	if (err) return 1;

/* FIXME: implement this ... */
#if 0
	if (!forceSanity) {
		/*
		 * Check if the device contains an ext2 filesystem
		 * ... tough with an LVM!
		 */
		if ((lvmseek(fd, 1)) == -1)
			return 1;
		if ((read(fd, buffer, LVM_BLOCKSIZE)) != LVM_BLOCKSIZE)
			return 1;
		if (buffer[56] == 0x53 && buffer[57] == 0xef && buffer[33] == 0x20 && buffer[37] == 0x20) {
			fprintf(stderr, "%s appears to contain an ext2 filesystem -- use -f to override\n", name);
			return 1;
		}
	}
#endif
	if (lvmseek(fd, 0) == -1)
		return 1;
	if ((read(fd, buffer, LVM_BLOCKSIZE)) != LVM_BLOCKSIZE)
		return 1;
	pv_sb = (pv_sb_t *) buffer;

	if (!forceSanity) {
		/*
		 * take a look at the superblock ...
		 */
		if (pv_sb->pv_magic == LVM_PV_SB_MAGIC) {
			fprintf(stderr, "PV %s appears to be already created -- use -f to\nforce the destruction of old contents\n", name);
			return 1;
		}
	}

	memset(buffer,0,LVM_BLOCKSIZE);
	memset(buffer,0x7f,(LVM_PV_SB_GENERIC_WORDS-17)*4);

	printf("creating PV ... not fully yet :)\n");
	pv_sb->pv_magic = LVM_PV_SB_MAGIC;

	if (ioctl(fd, BLKGETSIZE, (unsigned long)&nr_blocks) == -1) {
		fprintf(stderr, "couldn't get device size for %s -- %s\n", name, strerror(errno));
		close(fd);
		return 1;
	}
	nr_blocks /= LVM_BLOCKSIZE/512;

	printf("%s's size: %ld KB.\n", name, nr_blocks*LVM_BLOCKSIZE/1024);

	bg_size = TOTAL_BLOCKS_PER_BG;
	pv_sb->pv_bg_size = bg_size;

	 /* 1 left for the superblock */
	nr_blocks--;

	/*
	 * Rounding to block groups, no partial block groups allowed.
	 */
	nr_blocks = (nr_blocks / pv_sb->pv_bg_size) * pv_sb->pv_bg_size;
	printf("%s's rounded size: %ld KB.\n",
				name, nr_blocks*LVM_BLOCKSIZE/1024);

	pv_sb->pv_total_size = nr_blocks;
	block_groups = nr_blocks / pv_sb->pv_bg_size;
	pv_sb->pv_block_groups = block_groups;

	pv_sb->pv_first_free = 1;
	pv_sb->pv_first_used = 0;
	pv_sb->pv_blocks_left = nr_blocks;
	pv_sb->pv_block_size = LVM_BLOCKSIZE;
	pv_sb->pv_pptr_size = sizeof(pv_pptr_t);

	pv_sb->pv_uuid0 = get_random();
	pv_sb->pv_uuid1 = get_random();
	pv_sb->pv_uuid2 = get_random();
	pv_sb->pv_uuid3 = get_random();

	pv_sb->pv_ctime = (md_u32) time(NULL);

        pv_sb->pv_major = MKPV_MAJOR_VERSION;
        pv_sb->pv_minor = MKPV_MINOR_VERSION;
        pv_sb->pv_patch = MKPV_PATCHLEVEL_VERSION;

	/*
	 * Check wether any field was left uninitialized, accidentally:
	 */
	{
		unsigned int * buf = (unsigned int *) buffer;

		for (i = 0; i < (LVM_PV_SB_GENERIC_WORDS-17); i++) {
			if (buf[i] == 0x7f7f7f7f) {
				printf("PV sb word %d uninitialized!\n", i);
				return 1;
			}
		}
	}

	if (lvmseek(fd, 0) == -1)
		return 1;
	if ((write(fd, buffer, LVM_BLOCKSIZE)) != LVM_BLOCKSIZE)
		return 1;

	mkvg(fd, forceSanity);

	/*
	 * Initialize all the block groups:
	 */
	pv_bg = (pv_block_group_t *) buffer;

	memset(pv_bg, 0, LVM_BLOCKSIZE);

	printf("initializing block groups on %s:\n", name);
	for (i = 0; i < block_groups; i++) {
		if (lvmseek(fd, i*bg_size + 2) == -1)
			return 1;
		if ((write(fd, buffer, LVM_BLOCKSIZE)) != LVM_BLOCKSIZE)
			return 1;
		if (!(i&31))
			sync();
		printf("#%d\r", i); fflush(stdout);
	}

	return 0;
}

int main (int argc, char *argv[])
{
    FILE *fp = NULL;
    int exit_status=0;
    int version = 0, help = 0;
    char * configFile = RAID_CONFIG;
    int force_flag = 0;
    char ** args;
    struct md_version ver;
    poptContext optCon;
    int i;
    struct poptOption optionsTable[] = {
	{ "configfile", 'c', POPT_ARG_STRING, &configFile, 0 },
	{ "force", 'f', 0, &force_flag, 0 },
	{ "help", 'h', 0, &help, 0 },
	{ "version", 'V', 0, &version, 0 },
	{ NULL, 0, 0, NULL, 0 }
    } ;

    optCon = poptGetContext("mkpv", argc, (const char **)argv, optionsTable, 0);
    if ((i = poptGetNextOpt(optCon)) < -1) {
	fprintf(stderr, "%s: %s\n", 
		poptBadOption(optCon, POPT_BADOPTION_NOALIAS),
		poptStrerror(i));
	usage();
	return EXIT_FAILURE;
    }

    if (help) {
	usage();
	return EXIT_FAILURE;
    } else if (version) {
	printf("mkpv version %d.%d.%d\n", MKPV_MAJOR_VERSION,
			MKPV_MINOR_VERSION, MKPV_PATCHLEVEL_VERSION);
	return EXIT_VERSION;
    }

    fp = fopen(configFile, "r");
    if (fp == NULL) {
	fprintf(stderr, "Couldn't open %s -- %s\n", configFile, 
	        strerror(errno));
	goto abort;
    }

    srand((unsigned int) time(NULL));
    if (parse_config(fp))
	goto abort;

    args = (char **)poptGetArgs(optCon);
    if (!args) {
	fprintf(stderr, "nothing to do!\n");
	usage();
	return EXIT_FAILURE;
    }

    if (getMdVersion(&ver)) {
	fprintf(stderr, "cannot determine md version: %s\n", strerror(errno));
	return EXIT_FAILURE;
    }

    while (*args) {
	if (force_flag) {
		fprintf(stderr, 
		"DESTROYING the contents of %s in 5 seconds, Ctrl-C if unsure!\n", *args);
		sleep(5);
	}
	if (mkpv(*args, force_flag))
		goto abort;
	args++;
    }

    fclose(fp);
    return 0;

abort:
    fprintf(stderr, "mkpv: aborted\n");
    exit_status = 1;
    if (fp)
	fclose(fp);
    return exit_status;
}


